sources/kms/custom-key-store.go

```diff
@@ -0,0 +1,121 @@
+package kms
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/overmindtech/aws-source/sources"
+	"github.com/overmindtech/sdp-go"
+)
+
+func customKeyStoreOutputMapper(_ context.Context, _ *kms.Client, scope string, _ *kms.DescribeCustomKeyStoresInput, output *kms.DescribeCustomKeyStoresOutput) ([]*sdp.Item, error) {
+	items := make([]*sdp.Item, 0)
+
+	for _, customKeyStore := range output.CustomKeyStores {
+		attributes, err := sources.ToAttributesCase(customKeyStore, "tags")
+		if err != nil {
+			return nil, err
+		}
+
+		item := sdp.Item{
+			Type:            "kms-custom-key-store",
+			UniqueAttribute: "customKeyStoreId",
+			Attributes:      attributes,
+			Scope:           scope,
+		}
+
+		switch customKeyStore.ConnectionState {
+		case types.ConnectionStateTypeConnected:
+			item.Health = sdp.Health_HEALTH_OK.Enum()
+		case types.ConnectionStateTypeConnecting:
+			item.Health = sdp.Health_HEALTH_PENDING.Enum()
+		case types.ConnectionStateTypeDisconnected:
+			item.Health = sdp.Health_HEALTH_UNKNOWN.Enum()
+		case types.ConnectionStateTypeFailed:
+			item.Health = sdp.Health_HEALTH_ERROR.Enum()
+		case types.ConnectionStateTypeDisconnecting:
+			item.Health = sdp.Health_HEALTH_PENDING.Enum()
+		default:
+			return nil, &sdp.QueryError{
+				ErrorType:   sdp.QueryError_OTHER,
+				ErrorString: "unknown Connection State",
+			}
+		}
+
+		if customKeyStore.CloudHsmClusterId != nil {
+			// +overmind:link cloudhsmv2-cluster
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "cloudhsmv2-cluster",
+					Method: sdp.QueryMethod_GET,
+					Query:  *customKeyStore.CloudHsmClusterId,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// Changing the CloudHSM cluster will affect the custom key store
+					In: true,
+					// Updating the custom key store will not affect the CloudHSM cluster
+					Out: false,
+				},
+			})
+		}
+
+		// TODO: Activate this after enabling get vpc by name
+		/*
+			if customKeyStore.XksProxyConfiguration != nil &&
+				customKeyStore.XksProxyConfiguration.VpcEndpointServiceName != nil {
```

Comment:
Sorry one more thing that I've just noticed: I think this is referring to the result of the [`DescribeVpcEndpointServices`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2#Client.DescribeVpcEndpointServices) API call since these have a [`ServiceName`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2@v1.174.0/types#ServiceDetail), as opposed to the name of a VPC. Checking the CLI there is a `aws ec2 describe-vpc-endpoint-services` command so I think this is more likely to relate to be a `GET` request for a `ec2-vpc-endpoint-service` type (which doesn't yet exist) than an `ec2-vpc`

Reply:
https://github.com/overmindtech/aws-source/pull/571/commits/46eae9dca5240803f4ee14c5a69d44e4e69adc07

You are absolutely right.

I used the same page on AWS docs but overlooked the source name.